project/annotations/utils.py

```diff
@@ -163,14 +161,24 @@ def apply_alleviate(image_id, label_scores_all_points):
             alleviate_was_applied = True
 
     if alleviate_was_applied:
-        after_saving_annotations(img)
+        after_saving_points_or_annotations(img)
 
 
-def after_saving_annotations(image):
+def after_saving_points_or_annotations(image):
     """
-    Run this function after saving a batch of Annotations for an Image.
+    Run this function after creating, updating, or deleting Annotations for
```

Comment:
I do wish we could get rid of this. It's a tall order for a new developer to know to call this method. In a perfect world we'd add a post-save hook to the Annotation model to execute these operations on the relevant image. Did you look into that option? Or perhaps something to consider for next round?

Reply:
Ok, so I looked a bit more at your comment and at this code. It seems to me we could add this as a post-save hook on `image.save()`.  We could also add a smaller post-save hook to `annotations.save()` which does this part:
```
last_annotation = image.annotation_set.order_by('-annotation_date').first()
image.last_annotation = last_annotation
image.save()
```

We'd also leave the code in the image save post-hook, so the use-case you discuss:
```
image = Image.objects.get(pk=image_id)
# <update annotations here>
image.confirmed = True
image.save()
```
should be fine.

Reply:
Hmm, but if the Annotation post-save hook calls `image.save()`, then that means the Image post-save hook also runs as a result, right?

Looking at all the places where this PR calls `after_saving_points_or_annotations()`, I think some kind of post-save signal (or an overridden save() method, come to think of it) would work fine in most places. There are still two cases I'm concerned about though:

1. During the `annotation_tool` view, after applying Alleviate, the Image object is then passed to the `annotation_tool.html` template, where its last_annotation value gets read. If the last_annotation gets updated by a signal / custom save method on Annotation (instead of by an 'after saving' function call), then the template still reads the old value of last_annotation. Unless we remember to call `refresh_from_db()` on the Image.

   I agree that having to remember the 'after saving' call isn't great, but I think having to remember a 'refresh' call is probably harder than that.

2. During the `upload_annotations_ajax` view, we save some annotations, and then still have some Image fields to update after (point generation method, CPC content/filename). So there's still a danger of overwriting the old value of last_annotation, unless the annotation saving code is moved later, *after* all the other Image fields are updated. But putting the updates in the proper order seems like another thing that's more subtle to remember compared to an 'after saving' call.

Reply:
> Hmm, but if the Annotation post-save hook calls image.save(), then that means the Image post-save hook also runs as a result, right?

Good point! So all we need to do in the annotation post-save is calling `image.save()`. That's even easier. 

For you new cases, let's see.

1) In this case, the extent of the damage is that we show a stale value of `last_annotation`? That seems fine to me. Or are there more severe implications? Of course, we could also do a `refresh_from_db()` call to refresh the item in memory. The reason I'm advocating for making the "after saving" call a hook is that it updates the DB state, which is critical to keep updated.

2) I don't understand the issue here. In my suggestion the `image.save()` and `annotations.save()` post hooks would both update last_annotations field. So it shouldn't matter which order you do the saves. Can you write out the sequence of events you are worried about?

Reply:
> 1. In this case, the extent of the damage is that we show a stale value of `last_annotation`? That seems fine to me. Or are there more severe implications? Of course, we could also do a `refresh_from_db()` call to refresh the item in memory. The reason I'm advocating for making the "after saving" call a hook is that it updates the DB state, which is critical to keep updated.

It shows a stale value on the annotation tool page in the case that Alleviate was applied. We already don't make it very clear to the user when Alleviate was applied, so I think the stale value can be quite misleading.

But yes, it's not as critical as keeping the DB state updated. And the annotation tool IS a rather weird case, because it might be the only data-saving view which is a GET instead of a POST, and doesn't just redirect to a 'read-only' view after it's done saving. So I'd be okay with adding more weirdness with a `refresh_from_db()` if it's just this case. (I've been thinking annotation tool should be split into separate POST/GET views, which would make this moot, but I guess that's another task for later.)

> 2. I don't understand the issue here. In my suggestion the `image.save()` and `annotations.save()` post hooks would both update last_annotations field. So it shouldn't matter which order you do the saves. Can you write out the sequence of events you are worried about?

Oh, my bad. Yeah, I think you're right about this, and my brain just hadn't caught up to the idea yet of `image.save()` having a hook.

I guess the only other concern I can think of is performance. I can try saving 200 annotations in the annotation tool, thus calling the hook 200 times, and see how it compares to the old code.